src/server/http_server.rs

```diff
@@ -0,0 +1,141 @@
+// limitations under the License.
+
+use futures::future::{ok, FutureResult};
+use futures::sync::oneshot::{Receiver, Sender};
+use futures::{self, Future};
+use hyper::service::service_fn;
+use hyper::{self, Body, Method, Request, Response, Server, StatusCode};
+use tokio_threadpool::{Builder, ThreadPool};
+
+use std::net::SocketAddr;
+use std::str::FromStr;
+
+use super::Result;
+use util::metrics::dump;
+
+pub struct HttpServer {
+    thread_pool: ThreadPool,
+    tx: Sender<()>,
+    rx: Option<Receiver<()>>,
+    addr: Option<SocketAddr>,
+}
+
+impl HttpServer {
+    pub fn new(http_thread_pool_size: usize) -> Self {
+        let thread_pool = Builder::new()
+            .pool_size(http_thread_pool_size)
+            .name_prefix("http-server-")
+            .after_start(|| {
+                info!("HTTP server started");
+            })
+            .before_stop(|| {
+                info!("stopping HTTP server");
+            })
+            .build();
+        let (tx, rx) = futures::sync::oneshot::channel::<()>();
+        HttpServer {
+            thread_pool,
+            tx,
+            rx: Some(rx),
+            addr: None,
+        }
+    }
+
+    pub fn start(&mut self, http_addr: String) -> Result<()> {
+        let addr = SocketAddr::from_str(&http_addr)?;
+
+        // TODO: support TLS for HTTP server.
+        let builder = Server::try_bind(&addr)?;
+
+        // Create an HTTP service.
+        let service = |req: Request<Body>| -> FutureResult<Response<Body>, hyper::Error> {
+            let mut response = Response::new(Body::empty());
+
+            match (req.method(), req.uri().path()) {
+                (&Method::GET, "/metrics") => match dump() {
+                    Err(e) => {
+                        *response.status_mut() = StatusCode::INTERNAL_SERVER_ERROR;
+                        *response.body_mut() =
+                            Body::from(format!("failed to get metrics, error: {:?}", e));
+                    }
+                    Ok(buffer) => {
+                        *response.body_mut() = Body::from(buffer);
+                    }
+                },
+                (&Method::GET, "/status") => return ok(response),
+                _ => {
+                    *response.status_mut() = StatusCode::NOT_FOUND;
+                }
+            };
+            ok(response)
+        };
+
+        // Start to serve.
+        let server = builder.serve(move || service_fn(service));
+        self.addr = Some(server.local_addr());
+        let graceful = server
+            .with_graceful_shutdown(self.rx.take().unwrap())
+            .map_err(|e| error!("HTTP server error: {:?}", e));
+        self.thread_pool.spawn(graceful);
+        Ok(())
+    }
+
+    pub fn stop(self) {
+        let _ = self.tx.send(());
+        self.thread_pool
+            .shutdown_now()
+            .wait()
+            .unwrap_or_else(|e| error!("failed to stop HTTP server, error: {:?}", e));
```

Comment:
It's just wired for users to see something as "HTTP server". It is neither an item in the configurations nor a reasonable commonly used KV interface. Please change all occurrences to something like "Status server" or "Status service".